Adding pwfilesupport

stromp opened the pull request:
Pull request to add support to create a password file from a databag and manage mailbox users.

zuazo commented:
@stromp thanks for your contribution! :smiley: 

Please, see [CONTRIBUTING](https://github.com/zuazo/dovecot-cookbook/blob/master/CONTRIBUTING.md). I need you to improve the following before merging it:

- Add the `password.erb` template.
- Add [ChefSpec](https://github.com/sethvargo/chefspec/blob/master/README.md) or [KitchenCI](http://kitchen.ci/) tests.
- Update the documentation in the [README](https://github.com/zuazo/dovecot-cookbook/blob/master/README.md) and the [metadata](https://github.com/zuazo/dovecot-cookbook/blob/master/metadata.rb#L58).
- Fix Foodcritic and RuboCop offenses (`$ rake style`):

```
FC033: Missing template: /home/yunweb/chef-repo-onddo/cookbooks/dovecot/recipes/create_pwfile.rb:45
```

```
Offenses:

recipes/create_pwfile.rb:22:1: C: Extra blank line detected.
recipes/create_pwfile.rb:26:1: C: Extra blank line detected.
recipes/create_pwfile.rb:27:66: C: Space missing after comma.
passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
                                                                 ^
recipes/create_pwfile.rb:27:78: C: Surrounding space missing for operator |.
passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
                                                                             ^
recipes/create_pwfile.rb:27:81: C: Line is too long. [96/80]
passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
                                                                                ^^^^^^^^^^^^^^^^
recipes/create_pwfile.rb:27:91: C: Space missing after comma.
passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
                                                                                          ^
recipes/create_pwfile.rb:31:8: C: Space missing after comma.
  (user,crypt) = line.strip.split(":")
       ^
recipes/create_pwfile.rb:31:35: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
  (user,crypt) = line.strip.split(":")
                                  ^^^
recipes/create_pwfile.rb:37:1: C: Extra blank line detected.
recipes/create_pwfile.rb:38:46: C: Space missing after comma.
data_bag_item(node['dovecot']['databag_name'],node['dovecot']['databag_item_name'])['users'].each do |user|
                                             ^
recipes/create_pwfile.rb:38:81: C: Line is too long. [107/80]
data_bag_item(node['dovecot']['databag_name'],node['dovecot']['databag_item_name'])['users'].each do |user|
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
recipes/create_pwfile.rb:39:1: C: Use 2 (not 4) spaces for indentation.
    enc_password = shell_out("/usr/bin/doveadm pw -s MD5 -p #{user[1]}").stdout
^^^^
recipes/create_pwfile.rb:40:81: C: Line is too long. [126/80]
    credentials_updated = true if shell_out("/usr/bin/doveadm pw -t '#{local_creds[user[0]]}' -p #{user[1]}").exitstatus != 0 
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
recipes/create_pwfile.rb:40:126: C: Trailing whitespace detected.
    credentials_updated = true if shell_out("/usr/bin/doveadm pw -t '#{local_creds[user[0]]}' -p #{user[1]}").exitstatus != 0 
                                                                                                                             ^
recipes/create_pwfile.rb:44:1: C: Extra blank line detected.
recipes/create_pwfile.rb:46:10: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
  source "password.erb"
         ^^^^^^^^^^^^^^
recipes/create_pwfile.rb:50:12: W: (...) interpreted as grouped expression.
  variables ({
           ^
recipes/create_pwfile.rb:51:1: C: Tab detected.
        :credentials => credentials
^
recipes/create_pwfile.rb:51:2: C: Use the new Ruby 1.9 hash syntax.
        :credentials => credentials
 ^^^^^^^^^^^^^^^
recipes/create_pwfile.rb:51:2: C: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.
        :credentials => credentials
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
recipes/create_pwfile.rb:53:34: C: Trailing whitespace detected.
  only_if { credentials_updated } 
                                 ^
recipes/create_pwfile.rb:55:1: C: 4 trailing blank lines detected.

74 files inspected, 22 offenses detected
RuboCop failed!
```

Thanks again for your PR!!

stromp commented:
Hi Xavier,

Thanks for your reply!I did some fixing. However I get stuck on running the
tests.

when I run the unit tests I get this:

/opt/chefdk/embedded/bin/ruby
-I/home/stromp/.chefdk/gem/ruby/2.1.0/gems/rspec-core-3.4.1/lib:/home/stromp/.chefdk/gem/ruby/2.1.0/gems/rspec-support-3.4.1/lib
/home/stromp/.chefdk/gem/ruby/2.1.0/gems/rspec-core-3.4.1/exe/rspec
--pattern test/unit/\*\*\{,/\*/\*\*\}/\*_spec.rb --color --format progress
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/httpclient-2.6.0.1/lib/httpclient/webagent-cookie.rb:458:
warning: already initialized constant HTTPClient::CookieManager
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/httpclient-2.6.0.1/lib/httpclient/cookie.rb:8:
warning: previous definition of CookieManager was here

Randomized with seed 56860

Randomized with seed 56860
E, [2016-01-22T16:31:19.139154 #2302] ERROR -- : Actor crashed!
Faraday::Error: :gzip is not registered on Faraday::Response

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday.rb:189:in
`lookup_middleware'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:204:in
`use_symbol'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:96:in
`response'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/community_rest.rb:79:in
`block in initialize'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:66:in
`build'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/faraday-0.9.2/lib/faraday/rack_builder.rb:55:in
`initialize'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/community_rest.rb:77:in
`new'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/community_rest.rb:77:in
`initialize'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/downloader.rb:62:in
`new'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/downloader.rb:62:in
`try_download'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/downloader.rb:36:in
`block in download'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/downloader.rb:35:in
`each'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/downloader.rb:35:in
`download'

/home/stromp/.chefdk/gem/ruby/2.1.0/gems/berkshelf-3.3.0/lib/berkshelf/installer.rb:105:in
`install'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in
`public_send'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in
`dispatch'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/calls.rb:63:in
`dispatch'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:60:in
`block in invoke'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/cell.rb:71:in
`block in task'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/actor.rb:357:in
`block in task'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:57:in
`block in initialize'

/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:15:in
`block in create'
E, [2016-01-22T16:31:19.146592 #2302] ERROR -- : Actor crashed!

etc. etc

Is this something know to you? Or do you have an idea on how to progress
with this?

best regards,

Sjoerd Tromp


On Thu, Jan 21, 2016 at 11:30 PM, Xabier de Zuazo <notifications@github.com>
wrote:

> @stromp <https://github.com/stromp> thanks for your contribution! [image:
> :smiley:]
>
> Please, see CONTRIBUTING
> <https://github.com/zuazo/dovecot-cookbook/blob/master/CONTRIBUTING.md>.
> I need you to improve the following before merging it:
>
>    - Add the password.erb template.
>    - Add ChefSpec
>    <https://github.com/sethvargo/chefspec/blob/master/README.md> or
>    KitchenCI <http://kitchen.ci/> tests.
>    - Update the documentation in the README
>    <https://github.com/zuazo/dovecot-cookbook/blob/master/README.md> and
>    the metadata
>    <https://github.com/zuazo/dovecot-cookbook/blob/master/metadata.rb#L58>
>    .
>    - Fix Foodcritic and RuboCop offenses ($ rake style):
>
> FC033: Missing template: /home/yunweb/chef-repo-onddo/cookbooks/dovecot/recipes/create_pwfile.rb:45
>
> Offenses:
>
> recipes/create_pwfile.rb:22:1: C: Extra blank line detected.
> recipes/create_pwfile.rb:26:1: C: Extra blank line detected.
> recipes/create_pwfile.rb:27:66: C: Space missing after comma.
> passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
>                                                                  ^
> recipes/create_pwfile.rb:27:78: C: Surrounding space missing for operator |.
> passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
>                                                                              ^
> recipes/create_pwfile.rb:27:81: C: Line is too long. [96/80]
> passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
>                                                                                 ^^^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:27:91: C: Space missing after comma.
> passwordfile = File.open(node['dovecot']['conf']['password_file'],File::CREAT|File::RDONLY,0640)
>                                                                                           ^
> recipes/create_pwfile.rb:31:8: C: Space missing after comma.
>   (user,crypt) = line.strip.split(":")
>        ^
> recipes/create_pwfile.rb:31:35: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
>   (user,crypt) = line.strip.split(":")
>                                   ^^^
> recipes/create_pwfile.rb:37:1: C: Extra blank line detected.
> recipes/create_pwfile.rb:38:46: C: Space missing after comma.
> data_bag_item(node['dovecot']['databag_name'],node['dovecot']['databag_item_name'])['users'].each do |user|
>                                              ^
> recipes/create_pwfile.rb:38:81: C: Line is too long. [107/80]
> data_bag_item(node['dovecot']['databag_name'],node['dovecot']['databag_item_name'])['users'].each do |user|
>                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:39:1: C: Use 2 (not 4) spaces for indentation.
>     enc_password = shell_out("/usr/bin/doveadm pw -s MD5 -p #{user[1]}").stdout
> ^^^^
> recipes/create_pwfile.rb:40:81: C: Line is too long. [126/80]
>     credentials_updated = true if shell_out("/usr/bin/doveadm pw -t '#{local_creds[user[0]]}' -p #{user[1]}").exitstatus != 0
>                                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:40:126: C: Trailing whitespace detected.
>     credentials_updated = true if shell_out("/usr/bin/doveadm pw -t '#{local_creds[user[0]]}' -p #{user[1]}").exitstatus != 0
>                                                                                                                              ^
> recipes/create_pwfile.rb:44:1: C: Extra blank line detected.
> recipes/create_pwfile.rb:46:10: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
>   source "password.erb"
>          ^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:50:12: W: (...) interpreted as grouped expression.
>   variables ({
>            ^
> recipes/create_pwfile.rb:51:1: C: Tab detected.
>         :credentials => credentials
> ^
> recipes/create_pwfile.rb:51:2: C: Use the new Ruby 1.9 hash syntax.
>         :credentials => credentials
>  ^^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:51:2: C: Use 2 spaces for indentation in a hash, relative to the start of the line where the left curly brace is.
>         :credentials => credentials
>  ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> recipes/create_pwfile.rb:53:34: C: Trailing whitespace detected.
>   only_if { credentials_updated }
>                                  ^
> recipes/create_pwfile.rb:55:1: C: 4 trailing blank lines detected.
>
> 74 files inspected, 22 offenses detected
> RuboCop failed!
>
> Thanks again for your PR!!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zuazo/dovecot-cookbook/pull/17#issuecomment-173735739>
> .
>

zuazo commented:
Sorry for the delay. But it has taken me a while to fix the tests.

Please, could you rebase your branch onto the current master? I fixed the *faraday* error and all the integration tests.

svanharmelen commented:
@zuazo rebased and squashed...

zuazo commented:
@svanharmelen thank you.

There is a small RuboCop offense:

```
attributes/conf_files.rb:26:47: C: Don't use parentheses around a literal.
default['dovecot']['conf']['password_file'] = (
                                              ^
```

Could yo fix it to see if all the tests pass?

zuazo commented:
Please, could you look at the following things?

- Ensure there are no RuboCop offenses using [`$ rake style`](https://github.com/zuazo/dovecot-cookbook/blob/master/TESTING.md). There is a new one now:
```
attributes/conf_files.rb:26:81: C: Line is too long. [88/80]
default['dovecot']['conf']['password_file'] = "#{node['dovecot']['conf_path']}/password"
                                                                                ^^^^^^^^
```
- Add [ChefSpec](https://github.com/sethvargo/chefspec/blob/master/README.md) or [KitchenCI](http://kitchen.ci/) tests. Whatever you prefer.
- Add the new attributes to the README. Adding a small usage example is not necessary, but would also help.
- Add the new `recipe` to the *metadata.rb*

Of course, let me know if you need help with any of these things.

billiaz commented:
Hey Zuazo, I will continue also this pull request through Schubergphillis. Let me take a look on the history of it, and try to bring it up to date!

As this is a nice addition to the work I am doing for Dovecot.

zuazo commented:
@billiaz Thanks! Any help is really appreciated. Let me know if you have questions or need my help with it.

One problem I found with this PR, when I once tried to complete it myself, was that the `create_pwfile` recipe uses commands like `doveadm` at runtime. This will fail on first run because dovecot is still not installed during Chef compile time.

billiaz commented:
@zuazo 
please check this https://github.com/schubergphilis/dovecot-cookbook/tree/finalize_pwfilesupport

I am making changes under that branch and then merge them to the adding.... to avvoid triggering tests.

I am creating a depedency on other cookbooks / service installed for doveadm
Also created a migration from old style passwd:
user:pass to new one.

Lets see (I will run locally test kitchen for 1 of them and then merge the first changes).

Please if you have comments on my changes ping me!

billiaz commented:
Please close this request, I have submitted a new one which is based on another branch.

zuazo commented:
OK. Closing in favor of  #25.

zuazo commented:
Released in [`3.2.0`](https://github.com/zuazo/dovecot-cookbook/releases/tag/3.2.0).
